File: solutions/4081-maximum-transactions-without-negative-balance/maximum-transactions-without-negative-balance.java
import java.util.PriorityQueue;

public class Solution {

    public int maxTransactions(int[] transactions) {
        long valueOfTransactions = 0;
        int numberOfTransactions = 0;
        PriorityQueue<Integer> minHeapNegativeValuesIncludedInTransactions = new PriorityQueue<>();

        for (int value : transactions) {

            if (value >= 0) {
                ++numberOfTransactions;
                valueOfTransactions += value;
                continue;
            }

            if (valueOfTransactions + value >= 0) {
                minHeapNegativeValuesIncludedInTransactions.add(value);
                ++numberOfTransactions;
                valueOfTransactions += value;
                continue;
            }

            if (!minHeapNegativeValuesIncludedInTransactions.isEmpty() && minHeapNegativeValuesIncludedInTransactions.peek() < value) {
                valueOfTransactions += value - minHeapNegativeValuesIncludedInTransactions.poll();
                minHeapNegativeValuesIncludedInTransactions.add(value);
            }
        }

        return numberOfTransactions;
    }
}

File: solutions/4081-maximum-transactions-without-negative-balance/maximum-transactions-without-negative-balance.js
/**
 * @param {number[]} transactions
 * @return {number}
 */
var maxTransactions = function (transactions) {
    let valueOfTransactions = 0;
    let numberOfTransactions = 0;
    const minHeapNegativeValuesIncludedInTransactions = new PriorityQueue((x, y) => x - y);

    for (let value of transactions) {

        if (value >= 0) {
            ++numberOfTransactions;
            valueOfTransactions += value;
            continue;
        }

        if (valueOfTransactions + value >= 0) {
            minHeapNegativeValuesIncludedInTransactions.enqueue(value);
            ++numberOfTransactions;
            valueOfTransactions += value;
            continue;
        }

        if (!minHeapNegativeValuesIncludedInTransactions.isEmpty() && minHeapNegativeValuesIncludedInTransactions.front() < value) {
            valueOfTransactions += value - minHeapNegativeValuesIncludedInTransactions.dequeue();
            minHeapNegativeValuesIncludedInTransactions.enqueue(value);
        }
    }

    return numberOfTransactions;
};